zkevm-circuits/src/evm_circuit/execution/blockhash.rs

```diff
@@ -0,0 +1,201 @@
+            rw_counter: Delta(2.expr()),
+            program_counter: Delta(1.expr()),
+            gas_left: Delta(-OpcodeId::BLOCKHASH.constant_gas_cost().expr()),
+            ..Default::default()
+        };
+
+        let opcode = cb.query_cell();
+        let same_context = SameContextGadget::construct(cb, opcode, step_state_transition);
+        Self {
+            same_context,
+            block_number,
+            current_block_number,
+            block_hash,
+            block_lt,
+            diff_lt,
+        }
+    }
+
+    fn assign_exec_step(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        block: &Block<F>,
+        _: &Transaction,
+        _: &Call,
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        self.same_context.assign_exec_step(region, offset, step)?;
+
+        let block_number = block.rws[step.rw_indices[0]].stack_value();
+        self.block_number.assign(
+            region,
+            offset,
+            Some(
+                block_number.to_le_bytes()[..N_BYTES_U64]
+                    .try_into()
+                    .unwrap(),
+            ),
+        )?;
+        let block_number: F = block_number.to_scalar().unwrap();
+
+        let current_block_number = block.context.number;
+        self.current_block_number
+            .assign(region, offset, current_block_number.to_scalar())?;
+
+        self.block_hash.assign(
+            region,
+            offset,
+            Some(block.rws[step.rw_indices[1]].stack_value().to_le_bytes()),
+        )?;
+
+        self.block_lt.assign(
+            region,
+            offset,
+            block_number,
+            current_block_number.to_scalar().unwrap(),
+        )?;
+
+        self.diff_lt.assign(
+            region,
+            offset,
+            current_block_number.to_scalar().unwrap(),
+            block_number + F::from(256),
+        )?;
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::{
+        evm_circuit::witness::block_convert,
+        test_util::{test_circuits_using_witness_block, BytecodeTestConfig},
+    };
+    use bus_mapping::mock::BlockData;
+    use eth_types::{bytecode, geth_types::GethData, Bytecode, U256};
+    use mock::test_ctx::{helpers::*, TestContext};
+
+    fn test_ok(block_number: usize, current_block_number: u64) {
+        let mut code = Bytecode::default();
+        code.append(&bytecode! {
+            PUSH32(block_number)
+            BLOCKHASH
+            STOP
+        });
+
+        // simple U256 values for history hashes
+        let mut history_hashes = Vec::new();
+        if current_block_number < 256 {
+            for number in 0..current_block_number {
+                history_hashes.push(U256::from(number));
+            }
+        } else {
+            for number in (current_block_number - 256)..current_block_number {
+                history_hashes.push(U256::from(number));
+            }
+        }
```

Comment:
This should be easier to construct if you recieve a `Range` and also, the current block number + the number to lookup the hash for.

Will allow you to test failure cases for out of range things and similar stuff.

Reply:
Fixed. Is this way what you said?

Reply:
I don't see it changed TBH.
What I was referring is that you can get a Range struct which you can basically use to have the min and max values you need to iterate and push to the history hashes vec.

Reply:
Anyway, it's just a suggestion. Feel free to ignore.